### PR TITLE
feat: trim C0 control or space from input

### PR DESCRIPTION
**Pull request by anonrig**

### CODE REVIEW

File: src/parser.h [hunk 1]
```diff
@@ -21,6 +21,8 @@ namespace ada {
     bool password_token_seen = false;
 
     const char* pointer;
+    const char* pointer_start;
+    const char* pointer_end;
```
Comment (addaleax) [inline src/parser.h:25]: I don’t think it’s necessarily guaranteed that `std::string_view::iterator` is the same as `const char*`?

Comment (anonrig) [reply src/parser.h]: You're right. Updated the pull request. Thanks!

**Review by addaleax — APPROVED** [review]

File: src/parser.cpp [hunk 2]
```diff
@@ -110,7 +118,7 @@ namespace ada {
           // If url’s scheme is "file", then:
           if (url.scheme == "file") {
             // If remaining does not start with "//", validation error.
-            if (pointer + 2 < input.end() && pointer[1] == '/' && pointer[2] == '/') {
+            if (pointer + 2 < pointer_end && pointer[1] == '/' && pointer[2] == '/') {
```
Comment (lemire) [inline src/parser.cpp:121]: Consider using: `std::distance(pointer, pointer_end) < 2`.

Comment (lemire) [reply src/parser.cpp:121]: (Not a request for change.)

File: src/parser.cpp [hunk 3]
```diff
@@ -130,7 +138,7 @@ namespace ada {
           }
           // Otherwise, if remaining starts with an U+002F (/), set state to path or authority state
           // and increase pointer by 1.
-          else if (pointer + 1 < input.end() && pointer[1] == '/') {
+          else if (pointer + 1 < pointer_end && pointer[1] == '/') {
```
Comment (lemire) [inline src/parser.cpp:141]: same deal: `std::distance(pointer, pointer_end) < 1` or maybe `pointer != pointer_end` (would be faster if safe).

File: src/parser.cpp (hunk 2 above)
Comment (anonrig) [reply src/parser.cpp:121]: I didn't know `std::distance` existed. Thanks!

File: src/unicode.cpp [hunk 4]
```diff
@@ -27,4 +27,15 @@ namespace ada::unicode {
     return is_ascii_digit(c) || is_ascii_alpha(c);
   }
 
+  // A C0 control or space is a C0 control or U+0020 SPACE.
+  // A C0 control is a code point in the range U+0000 NULL to U+001F INFORMATION SEPARATOR ONE, inclusive.
+  bool is_c0_control_or_space(const char c) {
+    return c >= '\0' && c <= ' ';
```
Comment (lemire) [inline src/unicode.cpp:33]: This is fine, but note that it is equivalent to...
```
return (unsigned char) c <= ' ';
```

(Note that the compiler should know this as well.)

File: src/parser.cpp (hunk 2 above)
Comment (anonrig) [reply src/parser.cpp:121]: Upon research I saw that `std::distance` seems to be slower. https://stackoverflow.com/questions/57258226/stddistance-is-slow-and-how-to-improve-it

Comment (addaleax) [reply src/parser.cpp:121]: This seems to be a concern for non-random-access iterators like those used for maps only, though.

**Comment by lemire** [comment]
This does not obviously look correct to me... if it is, then it should be explained...

```C++
 bool is_ascii_digit(const char c) {
    return c >= 0 && c <= 9;
  }
```

Maybe you meant...


```C++
 bool is_ascii_digit(const char c) {
    return c >= '0' && c <= '9';
  }
```

File: src/parser.cpp (hunk 2 above)
Comment (lemire) [reply src/parser.cpp:121]: @addaleax is correct. It should never be slower than pointer arithmetic.

**Pull request merged**